_drafts/choosing_or_building_an_embedded_db.md

```diff
@@ -0,0 +1,330 @@
+struct which is read from a specific flash address on boot and written to that address
+when the user changes the settings:
+
+```c
+typedef struct {
+    int salinity;
+    int oxygen;
+} fishtank_config_t;
+
+fishtank_config_t config;
+
+void load() {
+  flash_read(CONFIG_ADDR, &config, sizeof(config));
+}
+
+void save() {
+  flash_erase_page(CONFIG_ADDR);
+  flash_program(CONFIG_ADDR, &config, sizeof(config));
+}
+```
+
+This solution may "just work" for an MVP fishtank,
+but as the product evolves we are going to run into a few issues:
+
+ 1. The persisted data is not portable due to endianness and integer size differences,
+ 2. The config will be corrupted if power is lost during the `save()` function,
+ 3. If a `temperature` field is added to the struct, it will be read as `-1` upon 
+    firmare upgrade, which may be unexpected.
+
+So we see that even for the simplest products, persisting data to flash
+presents some mantainability challenges.
+
+## Aside: Portability of Records
+
+Thinking ahead, you may eventually wish to run your fishtank code on 
+different hardware either because of an upgrade to the product,
+a new version of the product, or because you want to be able to load
+up records from the embedded system inside a mobile app or a server.
+
+That new processor may be 64-bit processor, or may even use a 
+different endianness than the microcontroller. Taking a closer look at
+the `fishtank_config_t` struct, we see that the basic `int` type is
+used, which uses 8 bytes on 64-bit versus 4 bytes on 32-bit,
+and the endianness (byte order) depends on the processor.
+
+### Integer Size and Alignment
+
+With the definition of `fishtank_config_t` above,
+`sizeof(fishtank_config_t)` will be different for different
+platforms. Compiling on a 64-bit MacBook where Apple
+kept `int` as 32-bit with 32-bit alignment, you'll get the 
+same 8-byte structure as on a 32-bit microcontroller.
+However, if you compile on an AWS EC2 instance
+(as you may want to do if there is a cloud component to 
+your fishtank), then you get a 16-byte structure because
+`int` there is 8 bytes.
+
+The bottom line for portability is, to use the definitions in `stdint.h`
+and instruct the compiler to pack your structure to remove any padding
+which would otherwise be added to encourage alignment to word boundries:
+
+```c
+#pragma pack(push, 1)
+typedef struct {
+    int32_t salinity;
+    int32_t oxygen;
+} fishtank_config_t;
+#pragma pack(pop)
+```
+
+### Much Ado About Endianness
+
+There's one remaining possible portability issue. There have historically 
+been "big-endian" processors which represent multi-byte integers with the
+most-significant byte first rather than least-significant byte first
+(so-called "little endian") which has become the dominant endianness today.
+This means that even with the fixed-sized, packed structure above,
+there may be processors where a record with `salinity == 2` is intepreted
+as `33554432` because of the endian mismatch.
```

Comment:
Might be worth a simple example of both to illustrate, though most of the audience should be familiar:

- Decimal value: 1234
- Hexadecimal: 0x04D2 (16-bit integer)
- Big-endian representation:

  |Address|Value|
  |---|---|
  |0x0|0x04 (LSB)|
  |0x1|0xD2 (MSB)|

- Little-endian representation:

  |Address|Value|
  |---|---|
  |0x0|0xD2 (MSB)|
  |0x1|0x04 (LSB)|

Reply:
Certainly. It's so easy to get it mixed up ;)

DONE (https://github.com/bondhome/interrupt/commit/9cfe15247c544a52f1bc95268e90ba0077175d05)